"panic: close of closed channel" in v2.2.1

### Welcome

- [x] Yes, I'm using a binary release within 2 latest [releases](https://github.com/golangci/golangci-lint/releases). Only such installations are supported.
- [x] Yes, I've searched similar [issues on GitHub](https://github.com/golangci/golangci-lint/issues) and didn't find any.
- [x] Yes, I've read the `typecheck` section of the [FAQ](https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors).
- [x] Yes, I've tried with the standalone [linter](https://golangci-lint.run/usage/linters/) if available (e.g., gocritic, go vet, etc.).
- [x] I agree to follow this project's [Code of Conduct](https://github.com/golangci/golangci-lint?tab=coc-ov-file)

### How did you install golangci-lint?

Nix

### Description of the problem

golangci-lint v2.2.1 just panicked for me on my https://github.com/leonklingele/lmk repo. I couldn't reproduce the panic, so it's probably a data race.

Steps to reproduce, including the config files & specific version of the binaries used:

```console
$ git clone --depth 1 https://github.com/leonklingele/lmk
$ cd lmk
$ git checkout d2e28797d45b1c7a290dd3079c054a2568b74f37
$ nix-shell # or direnv allow

# Apply the following patch:
diff --git a/main.go b/main.go
index 1e99350ac0db92349d3c565428dee2cc4187f170..5116a696ada57f032c80a6ef3c8e6ba54c7bd11a 100644
--- a/main.go
+++ b/main.go
@@ -427 +427 @@ func main() {
-    ctx := context.Background()
+    ctx = context.Background()

$ OUT="" ; until [[ "$OUT" == *"panic"* ]]; do OUT="$(make lint 2>&1)" ; done ; echo "$OUT"
INFO golangci-lint has version 2.2.1 built with go1.24.4 from v2.2.1 on 19700101-00:00:00
INFO [config_reader] Config search paths: [./ /Users/leon/code/lmk /Users/leon/code /Users/leon /Users /]
INFO [config_reader] Used config file .golangci.yml
INFO [config_reader] Module name "github.com/leonklingele/lmk"
INFO maxprocs: Leaving GOMAXPROCS=14: CPU quota undefined
INFO [goenv] Read go env for 4.600167ms: map[string]string{"GOCACHE":"/Users/leon/code/lmk/.gopath/.cache", "GOROOT":"/nix/store/rq7irijkj3nhapmjcv9d96xgkisj55x2-go-1.24.4/share/go"}
INFO [lintersdb] Active 89 linters: [asasalint asciicheck bidichk bodyclose canonicalheader containedctx contextcheck copyloopvar decorder depguard dogsled dupword durationcheck embeddedstructfieldcheck err113 errcheck errchkjson errname errorlint exhaustive exptostd fatcontext forbidigo forcetypeassert funcorder gci ginkgolinter gocheckcompilerdirectives gochecknoglobals gochecknoinits gochecksumtype goconst gocritic gofmt gofumpt goimports gomoddirectives goprintffuncname gosec gosmopolitan govet grouper iface ineffassign intrange loggercheck makezero mirror misspell mnd musttag nakedret nestif nilerr nilnesserr nilnil noctx nolintlint nonamedreturns nosprintfhostport paralleltest perfsprint predeclared promlinter protogetter reassign recvcheck revive rowserrcheck sloglint spancheck sqlclosecheck staticcheck tagalign tagliatelle testableexamples testifylint testpackage thelper tparallel unconvert unparam unused usestdlibvars usetesting wastedassign whitespace wrapcheck zerologlint]
INFO [loader] Go packages loading at mode 8767 (types_sizes|exports_file|imports|name|compiled_files|deps|files) took 192.021375ms
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 371.959µs
panic: close of closed channel

goroutine 5508 [running]:
github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*loadingPackage).analyze.func2(0x14003e2da80)
	github.com/golangci/golangci-lint/v2/pkg/goanalysis/runner_loadingpackage.go:109 +0x11c
created by github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*loadingPackage).analyze in goroutine 786
	github.com/golangci/golangci-lint/v2/pkg/goanalysis/runner_loadingpackage.go:91 +0x174
make: *** [Makefile:14: lint] Error 2
```

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 2.2.1 built with go1.24.4 from v2.2.1 on 19700101-00:00:00
```

</details>

### Configuration

<details>

```console
See https://github.com/leonklingele/lmk/blob/d2e28797d45b1c7a290dd3079c054a2568b74f37/.golangci.yml
```

</details>

### Go environment

<details>

```console
$ go version && go env
# paste output here
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
# See above, not reproducible
```

</details>

### A minimal reproducible example or link to a public repository

<details>

See above

</details>

### Validation

- [x] Yes, I've included all information above (version, config, etc.).

### Supporter

- [ ] I am a sponsor/backer through [GitHub](https://github.com/sponsors/golangci/) or [OpenCollective](https://opencollective.com/golangci-lint)

I'm able to reproduce the issue locally:

```console
$ mkdir empty
$ cd empty
$ golangci-lint run
ERRO [linters_context] typechecking error: pattern ./...: directory prefix . does not contain main module or its selected dependencies 
0 issues.
$ golangci-lint run
panic: close of closed channel

goroutine 200 [running]:
github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*loadingPackage).analyze.func2(0x14001b108c8)
        github.com/golangci/golangci-lint/v2/pkg/goanalysis/runner_loadingpackage.go:109 +0x11c
created by github.com/golangci/golangci-lint/v2/pkg/goanalysis.(*loadingPackage).analyze in goroutine 175
        github.com/golangci/golangci-lint/v2/pkg/goanalysis/runner_loadingpackage.go:91 +0x174
```

<details><summary>Go env</summary>
<p>

```console
$ go env
AR='ar'
CC='cc'
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_ENABLED='1'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
CXX='c++'
GCCGO='gccgo'
GO111MODULE=''
GOARCH='arm64'
GOARM64='v8.0'
GOAUTH='netrc'
GOBIN=''
GOCACHE='/Users/alexandear/Library/Caches/go-build'
GOCACHEPROG=''
GODEBUG=''
GOENV='/Users/alexandear/Library/Application Support/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFIPS140='off'
GOFLAGS=''
GOGCCFLAGS='-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -ffile-prefix-map=/var/folders/q0/5_z6pvw574z1c0zcrr2xvk0r0000gn/T/go-build3063022420=/tmp/go-build -gno-record-gcc-switches -fno-common'
GOHOSTARCH='arm64'
GOHOSTOS='darwin'
GOINSECURE=''
GOMOD='/dev/null'
GOMODCACHE='/Users/alexandear/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='darwin'
GOPATH='/Users/alexandear/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/opt/homebrew/Cellar/go/1.24.4/libexec'
GOSUMDB='sum.golang.org'
GOTELEMETRY='local'
GOTELEMETRYDIR='/Users/alexandear/Library/Application Support/go/telemetry'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/opt/homebrew/Cellar/go/1.24.4/libexec/pkg/tool/darwin_arm64'
GOVCS=''
GOVERSION='go1.24.4'
GOWORK=''
PKG_CONFIG='pkg-config'
```

</p>
</details>

--- a/go.mod
+++ b/go.mod
@@ -134,6 +134,7 @@ require (
 	go.augendre.info/fatcontext v0.8.0
 	go.uber.org/automaxprocs v1.6.0
 	golang.org/x/mod v0.26.0
+	golang.org/x/sync v0.15.0
 	golang.org/x/sys v0.34.0
 	golang.org/x/tools v0.34.0
 	gopkg.in/yaml.v3 v3.0.1
@@ -213,7 +214,6 @@ require (
 	go.uber.org/zap v1.24.0 // indirect
 	golang.org/x/exp v0.0.0-20240909161429-701f63a606c0 // indirect
 	golang.org/x/exp/typeparams v0.0.0-20250210185358-939b2ce775ac // indirect
-	golang.org/x/sync v0.15.0 // indirect
 	golang.org/x/text v0.26.0 // indirect
 	google.golang.org/protobuf v1.36.6 // indirect
 	gopkg.in/ini.v1 v1.67.0 // indirect

--- a/pkg/goanalysis/runner_action.go
+++ b/pkg/goanalysis/runner_action.go
@@ -34,6 +34,7 @@ func (act *action) waitUntilDependingAnalyzersWorked(ctx context.Context, stopCh
 		if dep.Package == act.Package {
 			select {
 			case <-stopChan:
+				return
 			case <-ctx.Done():
 				return
 			case <-dep.analysisDoneCh:

--- a/pkg/goanalysis/runner_action.go
+++ b/pkg/goanalysis/runner_action.go
@@ -1,6 +1,7 @@
 package goanalysis
 
 import (
+	"context"
 	"fmt"
 	"runtime/debug"
 
@@ -28,11 +29,12 @@ func (actAlloc *actionAllocator) alloc() *action {
 	return act
 }
 
-func (act *action) waitUntilDependingAnalyzersWorked(stopChan chan struct{}) {
+func (act *action) waitUntilDependingAnalyzersWorked(ctx context.Context, stopChan chan struct{}) {
 	for _, dep := range act.Deps {
 		if dep.Package == act.Package {
 			select {
 			case <-stopChan:
+			case <-ctx.Done():
 				return
 			case <-dep.analysisDoneCh:
 			}

--- a/pkg/goanalysis/runner_loadingpackage.go
+++ b/pkg/goanalysis/runner_loadingpackage.go
@@ -14,6 +14,7 @@ import (
 	"sync"
 	"sync/atomic"
 
+	"golang.org/x/sync/errgroup"
 	"golang.org/x/tools/go/gcexportdata"
 	"golang.org/x/tools/go/packages"
 
@@ -83,36 +84,28 @@ func (lp *loadingPackage) analyze(stopChan chan struct{}, loadMode LoadMode, loa
 		return
 	}
 
-	var actsWg sync.WaitGroup
-
-	actsWg.Add(len(lp.actions))
+	var actsWg errgroup.Group
 
 	for _, act := range lp.actions {
-		go func(act *action) {
-			defer actsWg.Done()
-
+		actsWg.Go(func() error {
 			act.waitUntilDependingAnalyzersWorked(stopChan)
 
 			select {
 			case <-stopChan:
-				return
+				return nil
 			default:
 			}
 
 			act.analyzeSafe()
 
-			select {
-			case <-stopChan:
-				return
-			default:
-				if act.Err != nil {
-					close(stopChan)
-				}
-			}
-		}(act)
+			return act.Err
+		})
 	}
 
-	actsWg.Wait()
+	err := actsWg.Wait()
+	if err != nil {
+		close(stopChan)
+	}
 }
 
 func (lp *loadingPackage) loadFromSource(loadMode LoadMode) error {

--- a/pkg/goanalysis/runner_loadingpackage.go
+++ b/pkg/goanalysis/runner_loadingpackage.go
@@ -1,6 +1,7 @@
 package goanalysis
 
 import (
+	"context"
 	"errors"
 	"fmt"
 	"go/ast"
@@ -84,14 +85,16 @@ func (lp *loadingPackage) analyze(stopChan chan struct{}, loadMode LoadMode, loa
 		return
 	}
 
-	var actsWg errgroup.Group
+	actsWg, ctx := errgroup.WithContext(context.Background())
 
 	for _, act := range lp.actions {
 		actsWg.Go(func() error {
-			act.waitUntilDependingAnalyzersWorked(stopChan)
+			act.waitUntilDependingAnalyzersWorked(ctx, stopChan)
 
 			select {
 			case <-stopChan:
+				return nil
+			case <-ctx.Done():
 				return nil
 			default:
 			}

